File: src/Workspace.Server/Domain.cs
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Text.Json.Serialization;

namespace Workspace.Server;

public static class JsonDefaults
{
    public static readonly JsonSerializerOptions Options = new(JsonSerializerDefaults.Web)
    {
        UnmappedMemberHandling = JsonUnmappedMemberHandling.Disallow,
        MaxDepth = 32
    };

    public static T Clone<T>(T value) =>
        JsonSerializer.Deserialize<T>(JsonSerializer.Serialize(value, Options), Options)
        ?? throw new InvalidDataException("Stored workspace data could not be read.");
}

[JsonUnmappedMemberHandling(JsonUnmappedMemberHandling.Disallow)]
public sealed record UiSpec(string Root, Dictionary<string, UiElement> Elements);

[JsonUnmappedMemberHandling(JsonUnmappedMemberHandling.Disallow)]
public sealed record UiElement(string Type, JsonObject Props, List<string> Children);

public sealed class WorkspaceDocument
{
    public required string Id { get; init; }
    public required string Title { get; set; }
    public required string Objective { get; init; }
    public required string Provider { get; init; }
    public string Status { get; set; } = "queued";
    public string Summary { get; set; } = "";
    public DateTimeOffset CreatedAt { get; init; } = DateTimeOffset.UtcNow;
    public DateTimeOffset UpdatedAt { get; set; } = DateTimeOffset.UtcNow;
    public int Revision { get; set; }
    public long EventSequence { get; set; }
    public string? ClarificationId { get; set; }
    public UiSpec? Clarification { get; set; }
    public Dictionary<string, string> Answers { get; set; } = [];
    public List<AgentState> Agents { get; set; } = [];
    public List<ArtifactState> Artifacts { get; set; } = [];
    public List<JobState> Jobs { get; set; } = [];
    public List<FeedbackState> Feedback { get; set; } = [];
    public List<HumanDecision> Decisions { get; set; } = [];
    public CanvasLayout Layout { get; set; } = new([], new(36, 36, 0.85));
}

public sealed class AgentState : IJsonOnDeserialized
{
    private bool _assigned = true;
    private bool _assignmentSpecified;

    public required string Id { get; init; }
    public required string Name { get; set; }
    public required string Role { get; set; }
    public bool Assigned
    {
        get => _assigned;
        set { _assigned = value; _assignmentSpecified = true; }
    }
    public string Status { get; set; } = "idle";
    public string Message { get; set; } = "Ready when you are.";
    public string? SessionId { get; set; }
    public string? ActiveJobId { get; set; }

    void IJsonOnDeserialized.OnDeserialized()
    {
        // The original app pre-created idle specialist placeholders before any recruitment.
        if (!_assignmentSpecified && Id != AgentPlanPolicy.CoordinatorId &&
            string.IsNullOrEmpty(SessionId) && Status is "idle" or "waiting")
            _assigned = false;
    }
}

public sealed class ArtifactState
{
    public required string Id { get; init; }
    public required string Title { get; set; }
    public int CurrentRevision { get; set; }
    public int? AcceptedRevision { get; set; }
    public List<ArtifactRevision> Revisions { get; set; } = [];
}

public sealed class ArtifactRevision
{
    public required int Revision { get; init; }
    public int? BaseRevision { get; init; }
    public required UiSpec Spec { get; init; }
    public string Status { get; set; } = "proposed";
    public required string Source { get; init; }
    public string? JobId { get; init; }
    public string? Review { get; init; }
    public bool Conflict { get; init; }
    public DateTimeOffset CreatedAt { get; init; } = DateTimeOffset.UtcNow;
}

public sealed class JobState
{
    public required string Id { get; init; }
    public required string Kind { get; init; }
    public required int Order { get; init; }
    public string Status { get; set; } = "queued";
    public string Message { get; set; } = "Waiting for an agent.";
    public string? FeedbackId { get; init; }
    public string? RetryOf { get; init; }
    public string? ErrorCode { get; set; }
    public DateTimeOffset CreatedAt { get; init; } = DateTimeOffset.UtcNow;
    public DateTimeOffset? StartedAt { get; set; }
    public DateTimeOffset? CompletedAt { get; set; }
    public Dictionary<string, int> BaseRevisions { get; set; } = [];
}

public sealed class FeedbackState
{
    public required string Id { get; init; }
    public required string ClientRequestId { get; init; }
    public required string JobId { get; set; }
    public required string Text { get; init; }
    public string? ArtifactId { get; init; }
    public string? ElementId { get; init; }
    public int? Revision { get; init; }
    public required string TargetLabel { get; init; }
    public string? QuotedText { get; init; }
    public string Status { get; set; } = "queued";
    public DateTimeOffset CreatedAt { get; init; } = DateTimeOffset.UtcNow;
}

public sealed record HumanDecision(
    string Id, string ArtifactId, int Revision, string Kind, DateTimeOffset CreatedAt);

public sealed record CanvasPosition(double X, double Y);
public sealed record CanvasViewport(double X, double Y, double Zoom);
public sealed record CanvasLayout(Dictionary<string, CanvasPosition> Positions, CanvasViewport Viewport);

public sealed record WorkspaceSummary(
    string Id, string Title, string Objective, string Provider, string Status,
    DateTimeOffset UpdatedAt, int ActiveJobs, int PendingFeedback, int PendingReviews, int ArtifactCount);

public sealed record WorkspaceEvent(
    long Sequence, string WorkspaceId, string Type, string Message,
    string? JobId, string? AgentId, DateTimeOffset CreatedAt);

public sealed record CreateWorkspaceRequest(string Objective, string Provider = "copilot");
public sealed record AnswerRequest(string ClarificationId, Dictionary<string, string> Answers);
public sealed record RefreshClarificationRequest(string ClarificationId);
public sealed record FeedbackRequest(
    string Text, string ClientRequestId, string? ArtifactId = null,
    string? ElementId = null, int? Revision = null);
public sealed record ReviewRequest(int Revision, string Decision);
public sealed record EditRequest(int Revision, string ElementId, string Text);
public sealed record ApiError(string Code, string Message);

public sealed class WorkspaceException(string code, string message, int statusCode = 400)
    : Exception(message)
{
    public string Code { get; } = code;
    public int StatusCode { get; } = statusCode;
}

File: web/src/domain.ts
import { z } from 'zod';

const Id = z.string().min(1).max(200);
const Timestamp = z.string().min(1);
export const ProviderId = z.enum(['copilot', 'demo']);
export type Provider = z.infer<typeof ProviderId>;
export const WorkspaceStatus = z.enum(['queued', 'working', 'needsInput', 'review', 'idle', 'error']);
export const JobStatus = z.enum(['queued', 'running', 'cancelling', 'completed', 'failed', 'cancelled', 'interrupted']);

export const BootstrapSchema = z.object({
  csrfToken: z.string().min(1), model: z.string(), reasoningEffort: z.string().nullable(),
  providers: z.array(z.object({ id: ProviderId, label: z.string() })), version: z.string(),
});
export type Bootstrap = z.infer<typeof BootstrapSchema>;
export const ProviderStatusSchema = z.object({
  id: z.string(), state: z.enum(['ready', 'setupRequired', 'unavailable', 'unsupported']),
  message: z.string(), model: z.string(), reasoningEffort: z.string().nullable(),
  sdkVersion: z.string().nullable().optional(), cliVersion: z.string().nullable().optional(),
  setupCommand: z.string().nullable().optional(),
});
export type ProviderStatus = z.infer<typeof ProviderStatusSchema>;
export const SummarySchema = z.object({
  id: Id, title: z.string(), objective: z.string(), provider: ProviderId, status: WorkspaceStatus,
  updatedAt: Timestamp, activeJobs: z.number().int(), pendingFeedback: z.number().int(),
  pendingReviews: z.number().int(), artifactCount: z.number().int(),
});
export type WorkspaceSummary = z.infer<typeof SummarySchema>;
export const PositionSchema = z.object({ x: z.number().finite(), y: z.number().finite() });
export const ViewportSchema = z.object({
  x: z.number().finite(), y: z.number().finite(), zoom: z.number().finite().positive(),
});
export const LayoutSchema = z.object({
  positions: z.record(z.string(), PositionSchema), viewport: ViewportSchema,
});
export type Point = z.infer<typeof PositionSchema>;
export type Viewport = z.infer<typeof ViewportSchema>;
export type CanvasLayout = z.infer<typeof LayoutSchema>;
export const AgentSchema = z.object({
  id: Id, name: z.string(), role: z.string(), assigned: z.boolean().default(true),
  status: z.enum(['idle', 'working', 'waiting', 'complete', 'error', 'cancelled']),
  message: z.string(), sessionId: z.string().nullable(), activeJobId: z.string().nullable(),
});
export type Agent = z.infer<typeof AgentSchema>;
export const RevisionSchema = z.object({
  revision: z.number().int(), baseRevision: z.number().int().nullable(),
  spec: z.unknown(), status: z.enum(['proposed', 'accepted', 'rejected', 'superseded']),
  source: z.string(), jobId: z.string().nullable(), review: z.string().nullable(),
  conflict: z.boolean(), createdAt: Timestamp,
});
export type ArtifactRevision = z.infer<typeof RevisionSchema>;
export const ArtifactSchema = z.object({
  id: Id, title: z.string(), currentRevision: z.number().int(),
  acceptedRevision: z.number().int().nullable(), revisions: z.array(RevisionSchema),
});
export type Artifact = z.infer<typeof ArtifactSchema>;
export const JobSchema = z.object({
  id: Id, kind: z.enum(['initial', 'answers', 'feedback', 'clarify']), order: z.number().int(), status: JobStatus,
  message: z.string(), feedbackId: z.string().nullable(), retryOf: z.string().nullable(),
  errorCode: z.string().nullable(), createdAt: Timestamp,
  startedAt: Timestamp.nullable(), completedAt: Timestamp.nullable(),
  baseRevisions: z.record(z.string(), z.number().int()),
});
export type Job = z.infer<typeof JobSchema>;
export const FeedbackSchema = z.object({
  id: Id, clientRequestId: z.string(), jobId: Id, text: z.string(),
  artifactId: z.string().nullable(), elementId: z.string().nullable(), revision: z.number().int().nullable(),
  targetLabel: z.string(), quotedText: z.string().nullable(),
  status: z.string(), createdAt: Timestamp,
});
export type Feedback = z.infer<typeof FeedbackSchema>;
export const DecisionSchema = z.object({
  id: Id, artifactId: Id, revision: z.number().int(), kind: z.string(), createdAt: Timestamp,
});
export const WorkspaceSchema = z.object({
  id: Id, title: z.string(), objective: z.string(), provider: ProviderId, status: WorkspaceStatus,
  summary: z.string(), createdAt: Timestamp, updatedAt: Timestamp,
  revision: z.number().int(), eventSequence: z.number().int().nonnegative(),
  clarificationId: z.string().nullable(), clarification: z.unknown().nullable(),
  answers: z.record(z.string(), z.string()), agents: z.array(AgentSchema),
  artifacts: z.array(ArtifactSchema), jobs: z.array(JobSchema), feedback: z.array(FeedbackSchema),
  decisions: z.array(DecisionSchema), layout: LayoutSchema,
});
export type Workspace = z.infer<typeof WorkspaceSchema>;
export const EventSchema = z.object({
  sequence: z.number().int().nonnegative(), workspaceId: Id, type: z.string(), message: z.string(),
  jobId: z.string().nullable(), agentId: z.string().nullable(), createdAt: Timestamp,
});
export type WorkspaceEvent = z.infer<typeof EventSchema>;
export interface Selection { artifactId: string; revision: number; elementId: string }
export interface FeedbackRequest {
  text: string; clientRequestId: string; artifactId?: string; elementId?: string; revision?: number;
}
export const activeJob = (job: Job) => ['queued', 'running', 'cancelling'].includes(job.status);
export function summarize(workspace: Workspace): WorkspaceSummary {
  return {
    id: workspace.id, title: workspace.title, objective: workspace.objective, provider: workspace.provider,
    status: workspace.status, updatedAt: workspace.updatedAt, activeJobs: workspace.jobs.filter(activeJob).length,
    pendingFeedback: workspace.feedback.filter((item) => ['queued', 'running', 'cancelling', 'review'].includes(item.status)).length,
    pendingReviews: workspace.artifacts.reduce((count, artifact) =>
      count + artifact.revisions.filter((revision) => revision.status === 'proposed').length, 0),
    artifactCount: workspace.artifacts.length,
  };
}

export function newerWorkspace(previous: Workspace | null, incoming: Workspace): Workspace {
  if (!previous || previous.id !== incoming.id) return incoming;
  return incoming.eventSequence >= previous.eventSequence && incoming.revision >= previous.revision
    ? incoming : previous;
}
